lib/std/jsonutils.nim

```diff
@@ -92,31 +90,27 @@ proc checkJsonImpl(cond: bool, condStr: string, msg = "") =
 template checkJson(cond: untyped, msg = "") =
   checkJsonImpl(cond, astToStr(cond), msg)
 
-template fromJsonFields(a, b, T, keys) =
+template fromJsonFields(a, b, T, keys, allowMissingFields) =
   checkJson b.kind == JObject, $(b.kind) # we could customize whether to allow JNull
   var num = 0
   for key, val in fieldPairs(a):
     num.inc
     when key notin keys:
       if b.hasKey key:
         fromJson(val, b[key])
-      else:
-        # we could customize to allow this
+      elif not allowMissingFields:
         checkJson false, $($T, key, b)
-  checkJson b.len == num, $(b.len, num, $T, b) # could customize
+  if not allowMissingFields:
```

Comment:
* need to document that `allowMissingFields` only refers to non-discriminant keys (and discriminant keys are still required for object variants)

* this not really correct now that you've introduced `allowMissingFields`;
we need to distinguish whether a key is in a but not and b and vice versa, and user should be able to allow either scenarios independently.

I suggest:
```nim
type Joptions = object
  allowMissingKeys
  allowExtraKeys
```
(keys makes it more obvious we're referring to the json key instead of the object field).

* the check `checkJson b.len == num, $(b.len, num, $T, b)` is now incorrect now that you're allowing `allowMissingKeys`; maybe something like:

```nim
for key, val in fieldPairs(a):
  ...
  if b.hasKey key:
    numMatched.inc
    ...

if not opt.allowExtraKeys:
  checkJson b.len <= numMatched, $(b.len, numMatched, $T, b)
```

Reply:
Here `numMatched` will be always smaller or equal than `b.len`. I think that the correct check is:
```nim
if opt.allowExtraKeys and opt.allowMissingKeys:
  discard
elif opt.allowExtraKeys:
  # This check is redundant because if here missing keys are not allowed, and
  # if `b.len < num` it will fail in the loop above but it is left for clarity.
  checkJson b.len >= num, $(b.len, num, $T, b)
elif opt.allowMissingKeys:
  checkJson b.len <= num, $(b.len, num, $T, b)
else:
  checkJson b.len == num, $(b.len, num, $T, b)
```

which is a little verbose and it can be written in shorter form as:
```nim
if not opt.allowExtraKeys and not opt.allowMissingKeys:
  checkJson b.len == num, $(b.len, num, $T, b)
elif not opt.allowExtraKeys and opt.allowMissingKeys:
  checkJson b.len <= num, $(b.len, num, $T, b)
```

but I preferred to write it in the first form because it seems clearer to me. I can change it if you prefer the second one.

Reply:
I wrote `Joptions` object from the point of view of the JSON objects. Said with other words, extra keys and missing keys refer to JSON object keys. It can be also written to refer to Nim object fields. I'm not sure which is better and I could change it if you like?

Reply:
let's make sure it's correct first; right now I still don't think it's correct for the reasons I gave in https://github.com/nim-lang/Nim/pull/15133#discussion_r463775084 ; consider this:

```nim
type A = object
  a1,a2,a3: int
j = {"a3": 1, "a4": 2} # pseudo-code
```
with allowMissingKeys=true, and allowExtraKeys = false, your test passes but should fail because of extra key `a4`.
That's why you need `numMatched`

Reply:
You can see the last version of this in the last commit. You were completely correct about `numMatched`. If you don't find another not working test we still have to decide whether the version with only two `if`s isn't better?

Reply:
> I wrote Joptions object from the point of view of the JSON objects.

ya i think that's fine (hence "keys" and not fields). No strong opinion but i think it's clear.

>  we still have to decide whether the version with only two ifs isn't better?

agnostic to that but there is 1 more concern: code bloat. how about:

```nim
var ok = if ... elif .... else: ...
checkJson ok, $(json.len, num, numMatched, $T, json)
```
which reduces bloat.
note that code bloat impacts performance to some extent (IR cache)

Reply:
Fixed.

Reply:
nice!